src/meta/expr.lean

```diff
@@ -338,6 +338,9 @@ e.is_constructor d.to_name ∨
   d.to_name.last ∈ ["below", "binduction_on", "brec_on", "cases_on", "dcases_on", "drec_on", "drec",
   "rec", "rec_on", "no_confusion", "no_confusion_type", "sizeof", "ibelow", "has_sizeof_inst"])
 
+meta def univ_levels (d : declaration) : list level :=
```

Comment:
doc string

Reply:
I think the name says pretty much all of it. What is missing?

Reply:
I agree that there's not much more to say.

On the other hand, mathlib is missing so much of the documentation is should have, that I think we should be aspiring to having linting / tests that enforce documentation, as a mechanism to improve documentation practice. That can only work if there really are doc-comments for everything, so in the meantime @robertylewis is acting like a friendly version of the linting tool. :-)

How about just
```suggestion
/-- Returns the list of universe levels of a declaration. -/
meta def univ_levels (d : declaration) : list level :=
```

Reply:
I understand but if we enforce this kind of standard indiscriminately, we create a lot of noise documentation. Comments are also code that need to be maintained and they need to pull their weight. 

This being said, I agree that it makes it hard to enforce automatically.

Reply:
I agree that almost all tactic defs should have documentation. You might think that `univ_levels` is self documenting, but what is the level of a declaration? Is it the definitional height? Why are there multiple levels? It's easy to think this is obvious if you are in the right head-space, but documentation is there to help people to find that head-space. In fact, I would argue that the list of universe levels isn't even a property of a declaration, so much as a property that all present kinds of declaration have. If notations were considered declarations then they wouldn't have a `univ_levels` field.

Reply:
The first part of your message is what I was looking for. What is not obvious in the definition? Next question: do we need to explain what a `level` is every time it is referenced? Do we need to explain why a declaration has universe parameters every time we mention `declaration`?

I think `declaration` needs to be well documented. After that, I don't think this definition adds much that is complicated. 

The second part of your comment is really about how `declaration` is defined. It's not really pertinent to this definition.

Reply:
> Do we need to explain why a declaration has universe parameters every time we mention declaration?

No, but the function that extracts the universe parameters from a declaration seems like a good place to explain what are the universe parameters of a declaration.

> The second part of your comment is really about how declaration is defined. It's not really pertinent to this definition.

I disagree. It's true that the current declarations share a lot of common fields, but the question is what they all have in common that puts them in the same type. To me, a declaration is a thing that has a name and is stored in the environment. As such, I see the name as an essential part of a declaration (and in my hypothetical, this would imply that notations have names), but universe parameters are something that only constants (defined or otherwise) can have. As they currently exist, `declaration`s only contain constants, and so they inherit the properties of a constant: a type, a list of universe parameters, and possibly a value. This breakdown should exist somewhere as documentation, perhaps on the type `declaration`, perhaps distributed on the projections.

Reply:
Looking more carefully at this definition, it isn't obvious how this differs from `univ_params` which has a similar name, and why you would need this version. Perhaps the documentation here can be used to differentiate the two and say something about the use-case.